File: app/src/main/java/com/example/punchpad2/NoteViewModel.java
package com.example.punchpad2;

import android.app.Application;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.annotation.NonNull;

import com.stratonotes.NoteEntity;
import com.stratonotes.FolderWithNotes;

import java.util.List;

public class NoteViewModel extends AndroidViewModel {

    private final NoteRepository repository;

    public LiveData<List<FolderWithNotes>> getFoldersWithPreviews() {
        return repository.getFoldersWithNotes();

    }

    public void insert(NoteEntity note) {
        repository.insert(note);
    }

    public void delete(NoteEntity note) {
        repository.delete(note);
    }

    public void update(NoteEntity note) {
        repository.update(note);
    }

    public void toggleFavorite(NoteEntity note, boolean favorite) {
        NoteEntity updated = new NoteEntity(
                note.getId(),
                note.getFolderId(),
                note.getContent(),
                note.getCreatedAt(),
                System.currentTimeMillis(),
                favorite,
                note.isHiddenFromMain(),
                note.isLarge(),
                note.isTrashed()
        );
        repository.update(updated);
    }

    public LiveData<List<NoteEntity>> getTrashedNotes() {
        return repository.getTrashedNotes();
    }

    public void restore(NoteEntity note) {
        NoteEntity updated = new NoteEntity(
                note.getId(),
                note.getFolderId(),
                note.getContent(),
                note.getCreatedAt(),
                System.currentTimeMillis(),
                note.isFavorite(),
                false,
                note.isLarge(),
                note.isTrashed()
        );
        repository.update(updated);
    }

    public void permanentlyDelete(NoteEntity note) {
        repository.delete(note);
    }

    public LiveData<List<NoteEntity>> getAllNotes() {
        return allNotes;
    }

    private final LiveData<List<NoteEntity>> allNotes;

    public NoteViewModel(@NonNull Application application) {
        super(application);
        repository = new NoteRepository(application);
        this.allNotes = repository.getAllNotes(); // Initialize here
    }




}

File: app/src/main/java/com/stratonotes/NoteViewModel.kt
package com.stratonotes

import android.app.Application
import android.util.Log
import androidx.lifecycle.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class NoteViewModel(application: Application) : AndroidViewModel(application) {

    private val noteRepository: NoteRepository = NoteRepository(AppDatabase.getDatabase(application))
    val allNotes: LiveData<List<NoteEntity>>
    val trashedNotes: LiveData<List<NoteEntity>>
    val foldersWithNotes: LiveData<List<FolderWithNotes>>

    init {
        allNotes = noteRepository.allNotes
        trashedNotes = noteRepository.trashedNotes
        foldersWithNotes = noteRepository.foldersWithNotes
    }

    fun insert(note: NoteEntity) {
        viewModelScope.launch {
            noteRepository.insert(note)
        }
    }

    fun ensureStratoNotesFolder() {
        viewModelScope.launch {
            val existingFolders = noteRepository.getAllFolders()
            if (existingFolders.none { it.name == "StratoNotes" }) {
                noteRepository.insertFolder(FolderEntity(name = "StratoNotes"))
                Log.d("NoteViewModel", "StratoNotes folder created.")
            } else {
                Log.d("NoteViewModel", "StratoNotes folder already exists.")
            }
        }
    }

    fun delete(note: NoteEntity) {
        viewModelScope.launch {
            noteRepository.delete(note)
        }
    }

    fun update(note: NoteEntity) {
        viewModelScope.launch {
            noteRepository.update(note)
        }
    }

    fun toggleFavorite(note: NoteEntity, favorite: Boolean) {
        val updated = note.copy(
            isFavorite = favorite,
            lastEdited = System.currentTimeMillis()
        )
        update(updated)
    }

    fun restore(note: NoteEntity) {
        val restored = note.copy(
            isHiddenFromMain = false,
            lastEdited = System.currentTimeMillis()
        )
        update(restored)
    }

    fun permanentlyDelete(note: NoteEntity) {
        delete(note)
    }

    fun getFoldersWithPreviews(): LiveData<List<FolderWithNotes>> = foldersWithNotes

    fun searchNotes(query: String): LiveData<List<NoteEntity>> {
        return noteRepository.searchNotes(query).asLiveData()
    }

    fun emptyTrash() {
        viewModelScope.launch(Dispatchers.IO) {
            noteRepository.permanentlyDeleteAllTrashedNotes()
        }
    }

    suspend fun getAllNotesNow(): List<NoteEntity> {
        return noteRepository.getAllNotesNow()
    }

    fun getTrashedContent(): LiveData<Pair<List<FolderWithNotes>, List<NoteEntity>>> {
        return MediatorLiveData<Pair<List<FolderWithNotes>, List<NoteEntity>>>().apply {
            var folders: List<FolderWithNotes>? = null
            var notes: List<NoteEntity>? = null

            val fSource = noteRepository.getTrashedFoldersWithNotesLive()
            val nSource = noteRepository.trashedNotes

            addSource(fSource) {
                folders = it
                if (notes != null) value = Pair(folders ?: emptyList(), notes ?: emptyList())
            }

            addSource(nSource) {
                notes = it.filter { note -> note.folderId == 0L }
                if (folders != null) value = Pair(folders ?: emptyList(), notes ?: emptyList())
            }
        }
    }
}
